Migrate api section to 6 0 (#11687)
storybookjs/storybook: docs/api/new-frameworks.md

---
title: 'Frameworks'
---

> **One of Storybook’s most powerful aspects is that it’s architected to support any web framework. It supports React, Vue, Angular, Web Components, Svelte and over a dozen others. This work in progress guide helps you get started on adding new framework support for Storybook.**

Reviewer: ```suggestion
Storybook is architected to support diverse web frameworks including React, Vue, Angular, Web Components, Svelte and over a dozen others. This guide helps you get started on adding new framework support for Storybook.
```

Author: Fixed with latest commit

The real meat of this file is the framework presets, and these are standard [Storybook presets](./addons#addon-presets) -- you can look at framework packages in the Storybook monorepo (e.g. [react](https://github.com/storybookjs/storybook/blob/next/app/react/src/server/options.ts), [vue](https://github.com/storybookjs/storybook/blob/next/app/vue/src/server/options.ts), [web-components](https://github.com/storybookjs/storybook/blob/next/app/web-components/src/server/options.ts)) to see examples of framework-specific customizations.

### Configuring the client

To configure the client, you must provide a framework specific render function. Before diving into the details, it’s important to understand how user-written stories relate to what is finally rendered on the screen.

#### Renderable objects

Storybook stories are ES6 functions that return a “renderable object.”

Consider the following React story:

```js
// Button.story.js
import { Button } from './Button';
export default { 
    title: 'Button', 
    component: Button 
}

export const Sample = () => (
  <Button label='hello button' />
);
```

In this case, the renderable object is the React element, `<Button .../>`.

> In most other frameworks, the renderable object is actually a plain old javascript object. 

Reviewer: ```suggestion
In most other frameworks, the renderable object is actually a plain JavaScript object. 
```

Author: Fixed with latest commit

  template: '<button label=:label />',
  data: {
      label: 'hello button'
    }
});
```

The design of this “renderable object” is framework-specific, and should ideally match the idioms of that framework.

#### Render function

The frameworks render function is the thing responsible for converting the renderable object into DOM nodes. This is typically of the form:

```js
const rootElement = document.getElementById('root');

export default function renderMain({
  storyFn,
}: RenderMainArgs) {
  const storyObj = storyFn();
  const html = fn(storyObj);
  rootElement.innerHTML = html;
}
```

#### Package structure

On the client side, the key file is [`src/client/preview.js`](../configure/overview#configure-story-rendering):

```js
import { start } from '@storybook/core/client';
import './globals';
import render from './render';
const api = start(render);

// the boilerplate code
```

The globals file typically sets up a single global variable that client-side code (such as addon-provided decorators) can refer to if needed to understand which framework its running in:

```js
import { window } from 'global';
window.STORYBOOK_ENV = ‘vue’;
```

<div style="background-color:#F8FAFC">

> TODO: need location on the render function for the frameworks

Reviewer 2: - https://github.com/storybookjs/storybook/blob/next/app/react/src/client/preview/render.tsx
- https://github.com/storybookjs/storybook/blob/next/app/angular/src/client/preview/render.ts
- https://github.com/storybookjs/storybook/blob/next/app/vue/src/client/preview/render.ts

et al.

Author: Thanks for the links. They were added. Marking this as resolved
